nifi-nar-bundles/nifi-graph-bundle/nifi-graph-processors/src/main/java/org/apache/nifi/processors/graph/ExecuteGraphQueryRecord.java

```diff
@@ -0,0 +1,265 @@
+    public static final Relationship REL_GRAPH = new Relationship.Builder().name("response").description("The response object from the graph server").build();
+
+    public static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+            REL_SUCCESS, REL_FAILURE, REL_ERRORS, REL_GRAPH
+    )));
+
+    public static final String RECORD_COUNT = "records.count";
+    public static final String GRAPH_OPERATION_TIME = "graph.operations.took";
+    private volatile RecordPathCache recordPathCache;
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return DESCRIPTORS;
+    }
+
+    private GraphClientService clientService;
+    private RecordReaderFactory recordReaderFactory;
+    private RecordSetWriterFactory recordSetWriterFactory;
+    private ObjectMapper mapper = new ObjectMapper();
+
+    @OnScheduled
+    public void onScheduled(ProcessContext context) {
+        clientService = context.getProperty(CLIENT_SERVICE).asControllerService(GraphClientService.class);
+        recordReaderFactory = context.getProperty(READER_SERVICE).asControllerService(RecordReaderFactory.class);
+        recordSetWriterFactory = context.getProperty(WRITER_SERVICE).asControllerService(RecordSetWriterFactory.class);
+        recordPathCache = new RecordPathCache(context.getProperties().size() * 100);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if ( input == null ) {
+            return;
+        }
+
+        FlowFile output = session.create(input);
+        boolean failed = false;
+        final AtomicLong errors = new AtomicLong();
+        long delta = 0;
+        Map<String, String> attributes = new HashMap<>();
+        List<Map<String,Object>> graphResponses = new ArrayList<>();
+        try (InputStream is = session.read(input);
+             OutputStream os = session.write(output);
+             RecordReader reader = recordReaderFactory.createRecordReader(attributes, is, -1l, getLogger());
+             RecordSetWriter writer = recordSetWriterFactory.createWriter(getLogger(), reader.getSchema(), os, input.getAttributes());
+        ) {
+            Record record;
+
+            long start = System.currentTimeMillis();
+            long recordIndex = 0;
+            long batchStart = start;
+            String recordScript = getScript(SUBMISSION_SCRIPT, context, input);
+            writer.beginRecordSet();
+            while ((record = reader.nextRecord()) != null) {
+                ConcurrentMap<String, List<Object>> dynamicPropertyMap = new ConcurrentHashMap<>();
+                for (PropertyDescriptor entry : context.getProperties().keySet()) {
+                    if(entry.isDynamic()) {
```

Comment:
It would be a good idea to cache the acquisition of the dynamic properties at the top of the whole loop with something like this:

```
List<PropertyDescriptor> dynamic = context.getProperties()
   .keySet().stream()
   .filter(desc -> desc.isDynamic()...
```

Reply:
Good call. Refactored.